flyteadmin/pkg/server/service.go

```diff
@@ -422,11 +426,35 @@ func serveGatewayInsecure(ctx context.Context, pluginRegistry *plugins.Registry,
 		ReadHeaderTimeout: time.Duration(cfg.ReadHeaderTimeoutSeconds) * time.Second,
 	}
 
-	err = server.ListenAndServe()
-	if err != nil {
-		return errors.Wrapf(err, "failed to Start HTTP Server")
+	go func() {
+		err = server.ListenAndServe()
+		if err != nil && err != http.ErrServerClosed {
+			logger.Fatalf(ctx, "Failed to start HTTP Server: %v", err)
+		}
+	}()
+
+	// Gracefully shutdown the servers
+	sigCh := make(chan os.Signal, 1)
+	signal.Notify(sigCh, syscall.SIGINT, syscall.SIGTERM)
+	<-sigCh
+
+	shutdownCtx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	if err := server.Shutdown(shutdownCtx); err != nil {
+		logger.Errorf(ctx, "Failed to shutdown HTTP server: %v", err)
 	}
 
+	grpcShutdownCtx, grpcCancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer grpcCancel()
+
+	go func() {
+		grpcServer.GracefulStop()
+		grpcCancel()
+	}()
+	<-grpcShutdownCtx.Done()
```

Comment:
Waiting for graceful shutdown doesn't necessarily mean that it completed successful graceful shutdown. We should follow up with a more forceful stop per the documentation/examples: https://github.com/grpc/grpc-go/blob/master/examples/features/gracefulstop/server/main.go#L93-L99